.doom.d/config.org
#+TITLE: My Configuration
#+AUTHOR:  Thanawat Techaumnuaiwit
* Introduction
 My Configuration for Doom! Feel free to copy but I probably copy pasted from
 somewhere else. :P
** Original Config information:
Keep for refrence later..
#+BEGIN_SRC elisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', where Emacs
;;   looks when you load packages with `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c g k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c g d') to jump to their definition and see how
;; they are implemented.
#+END_SRC
* Personal Info/Email setup
#+BEGIN_SRC elisp
(setq user-full-name "Thanawat Techaumnuaiwit")
(after! mu4e
  (setq!  mu4e-get-mail-command "mbsync -c ~/.config/mbsync/mbsyncrc -a"))
(set-email-account! "ucsb"
                    '((user-mail-address      . "thanawat@ucsb.edu")
                      (smtpmail-smtp-user     . "thanawat@ucsb.edu")
                      (mu4e-drafts-folder     . "/ucsb/[acc1].Drafts")
                      (mu4e-refile-folder     . "/ucsb/[acc1].Archive")
                      (mu4e-sent-folder       . "/ucsb/[acc1].Sent Mail")
                      (mu4e-trash-folder      . "/ucsb/[acc1].Trash")
                      (smtpmail-starttls-credentials . (("smtp.gmail.com" 587 nil nil)))
                      (smtpmail-default-smtp-server . "smtp.gmail.com")
                      (smtpmail-smtp-server . "smtp.gmail.com")
                      (smtpmail-smtp-service . 587)
                      (mu4e-compose-signature . "You can find me at https://thanawat.xyz\n---\nThanawat Techaumnuaiwit"))
                    t)

(set-email-account! "personal-gmail"
  '((mu4e-sent-folder       . "/personal-gmail/[acc2].Sent Mail")
    (mu4e-drafts-folder     . "/personal-gmail/[acc2].Drafts")
    (mu4e-trash-folder      . "/personal-gmail/[acc2].Trash")
    (mu4e-refile-folder     . "/personal-gmail/[acc2].Archive")
    (smtpmail-starttls-credentials . (("smtp.gmail.com" 587 nil nil)))
    (smtpmail-default-smtp-server . "smtp.gmail.com")
    (smtpmail-smtp-server . "smtp.gmail.com")
    (smtpmail-smtp-user     . "thanatechaumnuaiwit@gmail.com")
    (user-mail-address      . "thanatechaumnuaiwit@gmail.com")    ;; only needed for mu < 1.4
    (mu4e-compose-signature . "You can find me at https://thanawat.xyz\n---\nThanawat Techaumnuaiwit"))
  nil)
#+END_SRC

#+RESULTS:

* Aesthetics
** Font
#+BEGIN_SRC elisp
(setq doom-font (font-spec :family "Mononoki Nerd Font" :size 18))
(setq doom-font (font-spec :family "Hasklug Nerd Font Mono" :size 18))
(after! ispell
  (setq ispell-dictionary "en"))
#+END_SRC
** Line Numbers
#+BEGIN_SRC elisp
;; If you want to change the style of line numbers, change this to `relative' or
;; `nil' to disable it:
(setq display-line-numbers-type 'nil)
;; Makes visual-lines work better
(setq visual-fill-column-center-text t)
#+END_SRC
** Start emacs Fullscreen
#+BEGIN_SRC elisp
(add-to-list 'default-frame-alist '(fullscreen . maximized))
#+END_SRC
** Theme
#+BEGIN_SRC elisp
(setq doom-theme 'doom-wilmersdorf)
#+END_SRC
** Frame title
#+BEGIN_SRC elisp
;; explcitly set the frametitle because otherwise the frame title would show weird characters
;; https://www.emacswiki.org/emacs/FrameTitle
(setq frame-title-format "%b - Doom Emacs")
#+END_SRC
** Pop-up rules
#+begin_src emacs-lisp
(set-popup-rules!
  '(("^\\*info\\*" :slot 2 :side left :width 85 :quit nil)))
#+end_src
** Modeline
#+BEGIN_SRC emacs-lisp
(display-battery-mode)
#+END_SRC

* Evil
#+BEGIN_SRC elisp
(setq evil-escape-key-sequence "fd")
(map! :leader
      :desc "rss" "o s" #'=rss)
;;Make evil-mode up/down operate in screen lines instead of logical lines
;; (define-key evil-motion-state-map "j" 'evil-next-visual-line)
;; (define-key evil-motion-state-map "k" 'evil-previous-visual-line)
 ;;Also in visual mode
;; (define-key evil-visual-state-map "j" 'evil-next-visual-line)
;; (define-key evil-visual-state-map "k" 'evil-previous-visual-line)
#+END_SRC
* org
#+BEGIN_SRC elisp
(setq org-directory "~/org/")
(setq +org-capture-todo-file "agenda/todo.org")
(setq +org-capture-journal-file "agenda/journal.org")
(setq +org-capture-projects-file "agenda/projects.org")
(setq +org-capture-notes-file "agenda/notes.org")
(setq org-agenda-files (list (concat org-directory "agenda")))
#+END_SRC


** Org settings
#+BEGIN_SRC elisp
(after! org
  ;; TODO refactor!
  (setq org-capture-templates (append org-capture-templates
    '(("h" "Homework" entry (file "~/org/homework.org" ) "* TODO %?\n  %i\n") ("b" "Blog idea" entry (file "~/org/blog-ideas.org" ) "* TODO %?\n  %i\n"))))
  (require 'ox-extra)
  (ox-extras-activate '(ignore-headlines))
  (setq org-latex-listings 'minted
      org-latex-packages-alist '(("" "minted"))
      org-latex-pdf-process
      '("pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"
        "pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"))
  (setq org-re-reveal-root "/home/thanawat/reveal.js/")
  (add-to-list 'org-modules 'org-habit)
  )

#+END_SRC
** Anki
#+BEGIN_SRC elisp
(use-package! anki-editor
  :config
  (setq anki-editor-create-decks t))

(map! :localleader
      :map org-mode-map
      (:prefix ("k" . "Anki")
        :desc "Push" "p" 'anki-editor-push-notes
        :desc "Retry" "r" 'anki-editor-retry-failure-notes
        :desc "Insert" "n" 'anki-editor-insert-note
        (:prefix ("c" . "Cloze")
          :desc "Dwim" "d" 'anki-editor-cloze-dwim
          :desc "Region" "r" 'anki-editor-cloze-region
          )
        )
 )
#+END_SRC
** Mermaid
#+BEGIN_SRC elisp
;; (use-package! ob-mermaid
;;   :config
;; (setq ob-mermaid-cli-path "~/node_modules/.bin/mmdc"))
(use-package! mermaid-mode
  :mode "\\.mmd\\'"
  :config
  (setq mermaid-mmdc-location "~/custom_packages/node_modules/.bin/mmdc"))
#+END_SRC
** Org-roam
#+BEGIN_SRC elisp
;; (use-package! org-roam-server
;;   :ensure t
;;   :config
;;   (setq org-roam-server-host "127.0.0.1"
;;         org-roam-server-port 8080
;;         org-roam-server-export-inline-images t
;;         org-roam-server-authenticate nil
;;         org-roam-server-label-truncate t
;;         org-roam-server-label-truncate-length 60
        ;; org-roam-server-label-wrap-length 20))
#+END_SRC

** Org-journal
#+BEGIN_SRC elisp
(use-package org-journal
  :config
  (setq org-journal-file-type 'weekly)
  (setq org-journal-file-format "%Y-%m-%d.org")
  (setq org-journal-enable-agenda-integration t)
  )
(map! :leader
      :desc "New scheduled entry" "n j J" #'org-journal-new-scheduled-entry)

#+END_SRC
** Notifications
#+BEGIN_SRC elisp
(after! org
  (require 'appt)
  (require 'notifications)
  (setq appt-time-msg-list nil)    ;; clear existing appt list
  (setq appt-display-interval '5)  ;; warn every 5 minutes from t - appt-message-warning-time
  (setq
    appt-message-warning-time '15  ;; send first warning 15 minutes before appointment
    appt-display-mode-line nil     ;; don't show in the modeline
    appt-display-format 'window)   ;; pass warnings to the designated window function
  (setq appt-disp-window-function (function ct/appt-display-native))

  (appt-activate 1)                ;; activate appointment notification
  ; (display-time) ;; Clock in modeline
  (defun ct/appt-display-native (min-to-app new-time msg)
    (notifications-notify
           :title (format "Event in %s minutes" min-to-app) ; Title
           :body (format "%s" msg)
           :urgency 'normal
           ))
  ;; Agenda-to-appointent hooks
  (org-agenda-to-appt)             ;; generate the appt list from org agenda files on emacs launch
  (run-at-time "24:01" 3600 'org-agenda-to-appt)           ;; update appt list hourly
  (add-hook 'org-finalize-agenda-hook 'org-agenda-to-appt) ;; update appt list on agenda view
)
#+END_SRC
** Other mappings
#+BEGIN_SRC elisp
(map! :leader
      :prefix ("a" . "Personal Kbds")
      :desc "Add word to dictionary" "w" #'add-word-to-dictionary)
#+END_SRC
* Nov.el
#+BEGIN_SRC elisp
(use-package! nov
  :mode ("\\.epub\\'" . nov-mode)
  :hook (nov-mode . mixed-pitch-mode)
  :hook (nov-mode . visual-line-mode)
  :hook (nov-mode . visual-fill-column-mode)
  :config
  (setq nov-text-width t)
  (setq nov-variable-pitch nil))
#+END_SRC

* Rss(elfeed)
#+BEGIN_SRC elisp
(use-package! elfeed
  :config
    (setq elfeed-search-filter "@1-month-ago +unread +daily")
    (defun elfeed-v-mpv (url)
    "Watch a video from URL in MPV"
    (async-shell-command (format "mpv \"%s\"" url)))

    (defun elfeed-view-mpv (&optional use-generic-p)
    "Youtube-feed link"
    (interactive "P")
    (let ((entries (elfeed-search-selected)))
        (cl-loop for entry in entries
        do (elfeed-untag entry 'unread)
        when (elfeed-entry-link entry)
        do (elfeed-v-mpv it))
        (mapc #'elfeed-search-update-entry entries)
        (unless (use-region-p) (forward-line))))

    (define-key elfeed-search-mode-map (kbd "M-v") 'elfeed-view-mpv)
    (add-hook! 'elfeed-search-mode-hook 'elfeed-update)
    )
#+END_SRC
* Code
** lsp
#+BEGIN_SRC elisp
(after! lsp-ui
  (setq lsp-ui-sideline-show-hover t))
#+END_SRC

** c
#+BEGIN_SRC elisp
(after! cc-mode
  (setq c-basic-offset 2)
  (setq tab-width 2))

#+END_SRC
** python
#+BEGIN_SRC elisp
 (setq python-shell-interpreter "python3"
      flycheck-python-pycompile-executable "python3")
#+END_SRC
** color
#+BEGIN_SRC elisp
(add-hook! 'rainbow-mode-hook
(hl-line-mode (if rainbow-mode -1 +1)))
#+END_SRC
** haskell(hlint)
#+BEGIN_SRC emacs-lisp
(after! dante
  (add-to-list 'flycheck-disabled-checkers 'haskell-hlint))
#+END_SRC
* Direnv
#+BEGIN_SRC elisp
;; (use-package direnv
;;  :config
 ;; (direnv-mode))
#+END_SRC
* Presentations
#+BEGIN_SRC elisp
  (map! :map org-present-mode-keymap
        :g [C-right] #'org-present-next
        :g [C-left]  #'org-present-prev
        )
(after! org-tree-slide (setq org-tree-slide-never-touch-face t))
#+END_SRC
* Emojify
#+BEGIN_SRC elisp
(use-package emojify
  :hook (after-init . global-emojify-mode))
#+END_SRC
* Openwith
#+BEGIN_SRC emacs-lisp
;; Opens video file in mpv
;; using openwith for this is a kind of bloated solution, however it works
(use-package! openwith
  :after-call pre-command-hook
  :config
  (openwith-mode t)
  (setq openwith-associations '(("\\.mp4\\'" "mpv" (file)) ("\\.webm\\'" "mpv" (file))))
  )
#+END_SRC
* IRC
#+BEGIN_SRC emacs-lisp
(after! circe
(set-irc-server! "Freenode"
  `(
    :host "irc.freenode.net"
    :tls t
    :port 6697
    :nick "thiskappaisgrey"
    :nickserv-nick "thiskappaisgrey"
    :nickserv-password (lambda (&rest _) (+pass-get-secret "irc/thiskappaisgrey"))
    )))
#+END_SRC


.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', where Emacs
;;   looks when you load packages with `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c g k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c g d') to jump to their definition and see how
;; they are implemented.

(setq user-full-name "Thanawat Techaumnuaiwit")
(after! mu4e
  (setq!  mu4e-get-mail-command "mbsync -c ~/.config/mbsync/mbsyncrc -a"))
(set-email-account! "ucsb"
                    '((user-mail-address      . "thanawat@ucsb.edu")
                      (smtpmail-smtp-user     . "thanawat@ucsb.edu")
                      (mu4e-drafts-folder     . "/ucsb/[acc1].Drafts")
                      (mu4e-refile-folder     . "/ucsb/[acc1].Archive")
                      (mu4e-sent-folder       . "/ucsb/[acc1].Sent Mail")
                      (mu4e-trash-folder      . "/ucsb/[acc1].Trash")
                      (smtpmail-starttls-credentials . (("smtp.gmail.com" 587 nil nil)))
                      (smtpmail-default-smtp-server . "smtp.gmail.com")
                      (smtpmail-smtp-server . "smtp.gmail.com")
                      (smtpmail-smtp-service . 587)
                      (mu4e-compose-signature . "You can find me at https://thanawat.xyz\n---\nThanawat Techaumnuaiwit"))
                    t)

(set-email-account! "personal-gmail"
  '((mu4e-sent-folder       . "/personal-gmail/[acc2].Sent Mail")
    (mu4e-drafts-folder     . "/personal-gmail/[acc2].Drafts")
    (mu4e-trash-folder      . "/personal-gmail/[acc2].Trash")
    (mu4e-refile-folder     . "/personal-gmail/[acc2].Archive")
    (smtpmail-starttls-credentials . (("smtp.gmail.com" 587 nil nil)))
    (smtpmail-default-smtp-server . "smtp.gmail.com")
    (smtpmail-smtp-server . "smtp.gmail.com")
    (smtpmail-smtp-user     . "thanatechaumnuaiwit@gmail.com")
    (user-mail-address      . "thanatechaumnuaiwit@gmail.com")    ;; only needed for mu < 1.4
    (mu4e-compose-signature . "You can find me at https://thanawat.xyz\n---\nThanawat Techaumnuaiwit"))
  nil)

(setq doom-font (font-spec :family "Mononoki Nerd Font" :size 18))
(setq doom-font (font-spec :family "Hasklug Nerd Font Mono" :size 18))
(after! ispell
  (setq ispell-dictionary "en"))

;; If you want to change the style of line numbers, change this to `relative' or
;; `nil' to disable it:
(setq display-line-numbers-type 'nil)
;; Makes visual-lines work better
(setq visual-fill-column-center-text t)

(add-to-list 'default-frame-alist '(fullscreen . maximized))

(setq doom-theme 'doom-wilmersdorf)

;; explcitly set the frametitle because otherwise the frame title would show weird characters
;; https://www.emacswiki.org/emacs/FrameTitle
(setq frame-title-format "%b - Doom Emacs")

(setq evil-escape-key-sequence "fd")
(map! :leader
      :desc "rss" "o s" #'=rss)
;;Make evil-mode up/down operate in screen lines instead of logical lines
;; (define-key evil-motion-state-map "j" 'evil-next-visual-line)
;; (define-key evil-motion-state-map "k" 'evil-previous-visual-line)
 ;;Also in visual mode
;; (define-key evil-visual-state-map "j" 'evil-next-visual-line)
;; (define-key evil-visual-state-map "k" 'evil-previous-visual-line)

(setq org-directory "~/org/")
(setq +org-capture-todo-file "agenda/todo.org")
(setq +org-capture-journal-file "agenda/journal.org")
(setq +org-capture-projects-file "agenda/projects.org")
(setq +org-capture-notes-file "agenda/notes.org")
(setq org-agenda-files (list (concat org-directory "agenda")))

(after! org
  ;; TODO refactor!
  (setq org-capture-templates (append org-capture-templates
    '(("h" "Homework" entry (file "~/org/homework.org" ) "* TODO %?\n  %i\n") ("b" "Blog idea" entry (file "~/org/blog-ideas.org" ) "* TODO %?\n  %i\n"))))
  (require 'ox-extra)
  (ox-extras-activate '(ignore-headlines))
  (setq org-latex-listings 'minted
      org-latex-packages-alist '(("" "minted"))
      org-latex-pdf-process
      '("pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"
        "pdflatex -shell-escape -interaction nonstopmode -output-directory %o %f"))
  (setq org-re-reveal-root "/home/thanawat/reveal.js/")
  (add-to-list 'org-modules 'org-habit)
  )

(use-package! anki-editor
  :config
  (setq anki-editor-create-decks t))

(map! :localleader
      :map org-mode-map
      (:prefix ("k" . "Anki")
        :desc "Push" "p" 'anki-editor-push-notes
        :desc "Retry" "r" 'anki-editor-retry-failure-notes
        :desc "Insert" "n" 'anki-editor-insert-note
        (:prefix ("c" . "Cloze")
          :desc "Dwim" "d" 'anki-editor-cloze-dwim
          :desc "Region" "r" 'anki-editor-cloze-region
          )
        )
 )

;; (use-package! ob-mermaid
;;   :config
;; (setq ob-mermaid-cli-path "~/node_modules/.bin/mmdc"))
(use-package! mermaid-mode
  :mode "\\.mmd\\'"
  :config
  (setq mermaid-mmdc-location "~/custom_packages/node_modules/.bin/mmdc"))

;; (use-package! org-roam-server
;;   :ensure t
;;   :config
;;   (setq org-roam-server-host "127.0.0.1"
;;         org-roam-server-port 8080
;;         org-roam-server-export-inline-images t
;;         org-roam-server-authenticate nil
;;         org-roam-server-label-truncate t
;;         org-roam-server-label-truncate-length 60
        ;; org-roam-server-label-wrap-length 20))

(use-package org-journal
  :config
  (setq org-journal-file-type 'weekly)
  (setq org-journal-file-format "%Y-%m-%d.org")
  (setq org-journal-enable-agenda-integration t)
  )
(map! :leader
      :desc "New scheduled entry" "n j J" #'org-journal-new-scheduled-entry)

(after! org
  (require 'appt)
  (require 'notifications)
  (setq appt-time-msg-list nil)    ;; clear existing appt list
  (setq appt-display-interval '5)  ;; warn every 5 minutes from t - appt-message-warning-time
  (setq
    appt-message-warning-time '15  ;; send first warning 15 minutes before appointment
    appt-display-mode-line nil     ;; don't show in the modeline
    appt-display-format 'window)   ;; pass warnings to the designated window function
  (setq appt-disp-window-function (function ct/appt-display-native))

  (appt-activate 1)                ;; activate appointment notification
  ; (display-time) ;; Clock in modeline
  (defun ct/appt-display-native (min-to-app new-time msg)
    (notifications-notify
           :title (format "Event in %s minutes" min-to-app) ; Title
           :body (format "%s" msg)
           :urgency 'normal
           ))
  ;; Agenda-to-appointent hooks
  (org-agenda-to-appt)             ;; generate the appt list from org agenda files on emacs launch
  (run-at-time "24:01" 3600 'org-agenda-to-appt)           ;; update appt list hourly
  (add-hook 'org-finalize-agenda-hook 'org-agenda-to-appt) ;; update appt list on agenda view
)

(map! :leader
      :prefix ("a" . "Personal Kbds")
      :desc "Add word to dictionary" "w" #'add-word-to-dictionary)

(use-package! nov
  :mode ("\\.epub\\'" . nov-mode)
  :hook (nov-mode . mixed-pitch-mode)
  :hook (nov-mode . visual-line-mode)
  :hook (nov-mode . visual-fill-column-mode)
  :config
  (setq nov-text-width t)
  (setq nov-variable-pitch nil))

(use-package! elfeed
  :config
    (setq elfeed-search-filter "@1-month-ago +unread +daily")
    (defun elfeed-v-mpv (url)
    "Watch a video from URL in MPV"
    (async-shell-command (format "mpv \"%s\"" url)))

    (defun elfeed-view-mpv (&optional use-generic-p)
    "Youtube-feed link"
    (interactive "P")
    (let ((entries (elfeed-search-selected)))
        (cl-loop for entry in entries
        do (elfeed-untag entry 'unread)
        when (elfeed-entry-link entry)
        do (elfeed-v-mpv it))
        (mapc #'elfeed-search-update-entry entries)
        (unless (use-region-p) (forward-line))))

    (define-key elfeed-search-mode-map (kbd "M-v") 'elfeed-view-mpv)
    (add-hook! 'elfeed-search-mode-hook 'elfeed-update)
    )

(after! lsp-ui
  (setq lsp-ui-sideline-show-hover t))

(after! cc-mode
  (setq c-basic-offset 2)
  (setq tab-width 2))

(setq python-shell-interpreter "python3"
     flycheck-python-pycompile-executable "python3")

(add-hook! 'rainbow-mode-hook
(hl-line-mode (if rainbow-mode -1 +1)))

;; (use-package direnv
;;  :config
 ;; (direnv-mode))

(map! :map org-present-mode-keymap
        :g [C-right] #'org-present-next
        :g [C-left]  #'org-present-prev
        )
(after! org-tree-slide (setq org-tree-slide-never-touch-face t))

(use-package emojify
  :hook (after-init . global-emojify-mode))

(set-popup-rules!
  '(("^\\*info\\*" :slot 2 :side left :width 85 :quit nil)))

(display-battery-mode)

(after! dante
  (add-to-list 'flycheck-disabled-checkers 'haskell-hlint))

;; Opens video file in mpv
;; using openwith for this is a kind of bloated solution, however it works
(use-package! openwith
  :after-call pre-command-hook
  :config
  (openwith-mode t)
  (setq openwith-associations '(("\\.mp4\\'" "mpv" (file)) ("\\.webm\\'" "mpv" (file))))
  )

(after! circe
(set-irc-server! "Freenode"
  `(
    :host "irc.freenode.net"
    :tls t
    :port 6697
    :nick "thiskappaisgrey"
    :nickserv-nick "thiskappaisgrey"
    :nickserv-password (lambda (&rest _) (+pass-get-secret "irc/thiskappaisgrey"))
    )))
